### PR TITLE
Added pattern and mode saving

### PR DESCRIPTION
**Pull request by MattBoone78**
- added save / load. Save on button 0 press for now, load on startup
- renamed mode to omxMode to make it easier to search and be more descriptive
- added enum for mode, OMXMode
- changed switch statements for omxMode to use the enum values
- instead of statically initing the stepNoteP array, added initPatterns() which will populate it with the same data. Once we're loading patterns, we won't need to call this except on initial run
- fixed a case where a note param was checking > 0 instead of > -1 for the invalid / unset state
- set the mode select initial state to the current mode ( previously wasn't exposed because the current mode would always be MODE_MIDI )
- rolled stepType array into stepNoteP. can be saved as a block of StepNote structs
- save and load pattern length. initialize it to 16 steps if EEPROM is uninitialized
- recently added sequencer features are not yet saved ( direction, midi channel etc )

### CODE REVIEW

**Comment by szydek** [comment]
@MattBoone78  - definitely excited about these enhancements being included! During my testing I noticed that when choosing between modes (MI, S1, etc) there is a tendency to freeze (on OM for example) and leaves me unable to choose a mode. I have to power cycle. I'm assuming you haven't seen this behavior (?) - perhaps it's my library versions? I haven't had a chance to debug it much yet. Thanks for contributing!

Note: I did test compiling the main branch and not seeing this issue.

**Comment by MattBoone78** [comment]
Awesome. I’ll check out that crash. My first guess would be maybe you’re
cycling off the end, OM + 1 and I changed something so it doesn’t clamp
properly? I’ll be on tonight around 9pst and have a look. I was going to
save some of the other settings next ( midi, pattern direction )

File: config.h [hunk 1]
```diff
@@ -1,5 +1,24 @@
+enum OMXMode
+{
+     MODE_MIDI = 0,
+     MODE_S1,
+     MODE_S2,
+     MODE_OM,
 
-const int DEFAULT_MODE = 0;
+     NUM_OMX_MODES
```
Comment (szydek) [inline config.h:8]:
```suggestion
     NUM_OMX_MODES = 3
```

Comment (szydek) [reply config.h:8]: fyi - this appears to fix the issue I mentioned in my comment.

**Review by szydek — CHANGES_REQUESTED** [review]

**Comment by MattBoone78** [comment]
In the enum, NUM_OMX_MODES will be assigned the value 4 which is right. I
think the real bug is this line in OMX-27.ino, like 927

newmode = (OMXMode)constrain(newmode + amt, 0, NUM_OMX_MODES);

should be

newmode = (OMXMode)constrain(newmode + amt, 0, NUM_OMX_MODES - 1);

**Comment by MattBoone78** [comment]
Thanks for finding that @szydek

File: OMX-27.ino [hunk 2]
```diff
@@ -924,7 +924,7 @@ void loop() {
 			// set mode
 //			int modesize = NUM_OMX_MODES;
 //			Serial.println(modesize);
-	    	newmode = (OMXMode)constrain(newmode + amt, 0, NUM_OMX_MODES);
+	    	newmode = (OMXMode)constrain(newmode + amt, 0, NUM_OMX_MODES - 1);
```
Comment (szydek) [inline OMX-27.ino:927]: Ah yes, this makes sense. I tested this version this morning and appears to work as expected 👍

**Review by szydek — APPROVED** [review]
I vote to merge. Really nice being able to save/load patterns!

File: config.h (hunk 1 above)
Comment (szydek) [reply config.h:8]: Feel free to reject/delete my previous suggestion in favor of:
```
newmode = (OMXMode)constrain(newmode + amt, 0, NUM_OMX_MODES - 1);

File: OMX-27.ino [hunk 3]
```diff
@@ -191,33 +190,35 @@ void readPotentimeters(){
...
-				case 1: // SEQ1
+				case MODE_S1: // SEQ1
 					if (noteSelect && noteSelection){ // note selection - do P-Locks
 						potNum = k;
 						potCC = pots[k];
 						potVal = analogValues[k];
 								// stepNoteP[8] {notenum,vel,len,p1,p2,p3,p4,p5}
 						if (k < 4){ // only store p-lock value for first 4 knobs
-							stepNoteP[playingPattern][selectedStep][k+3] = analogValues[k];
+							stepNoteP[playing][selectedStep].params[k] = analogValues[k];
```
Comment (okyeron) [inline OMX-27.ino:210]: Bug? `playing` is a global for transport. 

This should be `playingPattern` perhaps?

Comment (MattBoone78) [reply OMX-27.ino:210]: Yup think you're right

**Comment by armandvedel** [comment]
In clearPattern and resetPatternDefaults, the default .len should be 0 instead of 1, right? (as length is now 0-15)

**Comment by MattBoone78** [comment]
Yes, that's correct. good catch!

**Comment by okyeron** [comment]
reverted my tests withSave UI from last night and did something more simple.

The AUX key will only save eeprom if you are in `enc_edit`. So long press encoder to get the mode menu then press AUX to save.

Also added some IFDEFs for changing pin configurations between DEV (my own unit), released (1.0) and MIDI-only versions of the hardware.

**Pull request merged**